### PR TITLE
[RFC] Session Fields

### PR DESCRIPTION
**Pull request by dainperkins**
New RFC for Session fields provided.

### CODE REVIEW

File: CHANGELOG.next.md [hunk 1]
```diff
@@ -61,6 +61,8 @@ Thanks, you're awesome :-) -->
 
 #### Added
 
+ * Added RFC for session fields #879
```
Comment (ebeahan) [inline CHANGELOG.next.md:64]: Perhaps holding off on `CHANGELOG` entries for RFCs would be best until field definitions adds/changes would be committed later (stage 2)?

Any thoughts @webmat @epixa?

File: rfcs/0000-rfc-session.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+## Usage
+
+Session fields are used to describe the sesison attributes of:
+ - Cleint VPN Sessions
+ - Network to Network VPN Sessions
+ - Network Access Sessions (NAC, WPA, EAP, etc.)
+ - Local or remote device login sessions (RDP, ICA, xWindows)
+ - administrative sessions on infrastructure devices
+ - administrative sessions on cloud or application management portals
+ - applications sessions (e.g. sql server odbc session, application acces session)
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:95]:
```suggestion
 - applications sessions (e.g. sql server odbc session, application access session)
```

File: rfcs/0000-rfc-session.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+-->
+
+<!--
+Stage 3: Add or update all remaining field definitions. The list should now be exhaustive. The goal here is to validate the technical details of all remaining fields and to provide a basis for releasing these field definitions as beta in the schema. Use GitHub code blocks with yml syntax formatting.
+-->
+
+## Usage
+
+Session fields are used to describe the sesison attributes of:
+ - Cleint VPN Sessions
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:89]:
```suggestion
 - Client VPN Sessions
```

File: rfcs/0000-rfc-session.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+Stage 2: Included a real world example source document. Ideally this example comes from the source(s) identified in stage 1. If not, it should replace them. The goal here is to validate the utility of these field changes in the context of a real world example. Format with the source name as a ### header and the example document in a GitHub code block with json formatting.
+-->
+
+<!--
+Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
+-->
+
+## Scope of impact
+
+Scope of impact should be minimal, with no breaking changes exepcted.
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:149]:
```suggestion
Scope of impact should be minimal, with no breaking changes expected.
```

File: CHANGELOG.next.md (hunk 1 above)
Comment (ebeahan) [reply CHANGELOG.next.md:64]: Discussed with @webmat - agreed the CHANGELOG entry isn't necessary until a later stage.

File: rfcs/0000-rfc-session.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+## Fields
+This RFC calls for the addition of session fields to describe events related to various types of "sessions" reported by appliances, security devices, systems, management portals, applications, etc.
+
+||
+|session.kind:            | local, remote, network
+|session.authorization:   | user, admin, service
+|session.type:            | system, virtual, application, wired, wireless, vpn
+|session.name             | locally relevant name if available (e.g. HQ Client VPN, Win19-VDI, FIN-EXCEL-vApp)
+|session.id               | session id provided by server or custom fingerprint
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:19]:
```suggestion
| Field | Description |
| ----- | ----------- |
|session.kind:            | local, remote, network
|session.authorization:   | user, admin, service
|session.type:            | system, virtual, application, wired, wireless, vpn
|session.name             | locally relevant name if available (e.g. HQ Client VPN, Win19-VDI, FIN-EXCEL-vApp)
|session.id               | session id provided by server or custom fingerprint
```

Cleans up the table formatting

File: rfcs/0000-rfc-session.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+
+
+## Fields (yaml)
+---
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:24]: Let's add code fencing for the YAML field section.

File: rfcs/0000-rfc-session.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
+# 0000: Session
+<!--^ The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC, taking care not to conflict with other RFCs.-->
+
+- Stage: 0 **0 (strawperson)** <!-- Update to reflect target stage -->
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:4]:
```suggestion
- Stage: **0 (strawperson)** <!-- Update to reflect target stage -->
```

File: rfcs/0000-rfc-session.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+  - - <166>Feb 03 2020 11:27:05 5508x-1_9.12(3): %ASA-6-605005: Login permitted from 192.168.1.250/59277 to management:192.168.1.10/ssh for user "JohnDoe"
+- Session End
+ - - <166>Feb 03 2020 11:27:05 5508x-1_9.12(3): %ASA-6-315011: SSH session from 192.168.1.250 on interface management for user JohnDoe disconnected by SSH server, reason: timeout
+
+Example 3: ASA Web VPN
+- Session Start
+- - <166>Feb 03 2020 11:27:05 5508x-1_9.12(3): %ASA-6-721016: WebVPN session for client user JohnDoe , 192.168.1.100 has been created.
+
+- Session End:
+- - <166>Feb 03 2020 11:27:05 5508x-1_9.12(3):%ASA-6-721018: WebVPN session for client user JohnDoe , IP 192.168.1.100 has been deleted.
```
Comment (ebeahan) [inline rfcs/0000-rfc-session.md:133]:
```suggestion
* Note, while there is an association id (session.id) created prior to wpa/802.1x authentication, building the session event from the eap success message allows for easier integration
 of fields like username, client.ip, etc. in an 802.1x or WPA environment

    * Base 802.11 Association:  (802.11 session start)
        * <134>1 1380653443.857790533 MR18 events type=association radio='1' vap='1' channel='2' rssi='23' aid='1687088497’

    * Base 802.11 Deassociation Message  (802.11 session end)
        * 1380653443.857790533 my_AP events type=disassociation radio='1' vap='2' channel='6' reason='8' instigator='2' duration='11979.728000' auth_neg_dur='1380653443.85779053324000' last_auth_ago='5.074000' is_wpa='1' full_conn='1.597000' ip_resp='1.597000' ip_src='192.168.111.251' arp_resp='1.265000' arp_src='192.168.111.251' dns_server='192.168.111.1' dns_req_rtt='1380653443.85779053335000' dns_resp='1.316000' aid='1813578850'


Example 2: ASA Admin Login
* Session start
    * <166>Feb 03 2020 11:27:05 5508x-1_9.12(3): %ASA-6-605005: Login permitted from 192.168.1.250/59277 to management:192.168.1.10/ssh for user "JohnDoe"
* Session End
    * <166>Feb 03 2020 11:27:05 5508x-1_9.12(3): %ASA-6-315011: SSH session from 192.168.1.250 on interface management for user JohnDoe disconnected by SSH server, reason: timeout

Example 3: ASA Web VPN
* Session Start
    * <166>Feb 03 2020 11:27:05 5508x-1_9.12(3): %ASA-6-721016: WebVPN session for client user JohnDoe , 192.168.1.100 has been created.

* Session End:
    * <166>Feb 03 2020 11:27:05 5508x-1_9.12(3):%ASA-6-721018: WebVPN session for client user JohnDoe , IP 192.168.1.100 has been deleted.
```

Markdown is being markdown and doing strange things with these bullets. I tried clearing it up a bit.

**Comment by ebeahan** [comment]
@dainperkins I pushed a couple of changes to clean up the formatting a bit. No content changes 😄

**Comment by ebeahan** [comment]
Phase 0 (Strawperson) Criteria:

- [ ] Discuss with domain or subject matter experts the utility of these changes
- [ ] Discuss with ECS team whether these changes seem appropriate for ECS

**Comment by ebeahan** [comment]
Again thanks @dainperkins for submitting. I believe adding `session` as a top-level field set does address a current gap in ECS. 

The documentation will need to be clear that this field set is intended for; I could see possible end-user confusion over whether `session` is intended for network session (which the `client`/`server` field sets are intended) vs. login sessions. I think there's also good opportunity with `session` to consider field reuse possibilities (e.g. nesting `user`).

**Comment by ebeahan** [comment]
Ping @webmat @MikePaquette @jamiehynds @jonathan-buttner @leehinman to get some ECS committee visibility and discussion started 😄

File: rfcs/0000-rfc-session.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+      level: extended
+      type: Session Name
+      description: The name field is meant to contain a locally significant identifier for the session as configured. This could represent a VPN group name, a wireless network name (ssid), a wired network segment, VDI service name, or application identifier.
+
+      example: HQ-Wireless
+
+    - name: id
+      level: extended
+      type: Session id
+      description: The id field is meant to contain a locall significant identifier for the session as provided by the observer or host reporting the session.  If no id is provided this field can remain blank, or a hash function similar to network.community_id can be used to discretely identify sessions from unique values.
```
Comment (bkimmel) [inline rfcs/0000-rfc-session.md:73]: :grey_question: locall -> locally

File: rfcs/0000-rfc-session.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+<!--
+Stage 3: Document resolutions for all existing concerns. Any new concerns should be documented along with their resolution. The goal here is to eliminate the risk of churn and instability by resolving outstanding concerns.
+-->
+
+<!--
+Stage 4: Document any new concerns and their resolution. The goal here is to eliminate risk of churn and instability by ensuring all concerns have been addressed.
+-->
+
+## Real-world implementations
+Session fields would allow for the further normalization of VPN logs, application logs(e.g. ftp logs from firewalls, SQL Server sessions) administrative sessions (search for admin sessions on non encrypted ports), analyze and track user session behaviors across numerous infrastructure and application log sources.
```
Comment (bkimmel) [inline rfcs/0000-rfc-session.md:183]: :grey_question: application logs(e.g. ftp logs from firewalls, SQL Server sessions) administrative -> application logs (e.g. ftp logs from firewalls, SQL Server sessions), administrative

Missing space and comma

File: rfcs/0000-rfc-session.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+## Fields (yaml)
+```yaml
+---
+- name: session
+  title: Session
+  group: 2
+  short: User, admin, application, network, or service sessions
+  description: |-
+    These fields contain information about various types of user sessions typically reported & logged in an enterprise.
```
Comment (bkimmel) [inline rfcs/0000-rfc-session.md:32]: :grey_question: This reads a little bit like a tautology ( "A session is something that is typically a session"  ).  Adding a more formal definition like "A session typically encompasses the data that a system uses to manage instances of authentication against user identities, roles, etc."

Comment (dainperkins) [reply rfcs/0000-rfc-session.md:32]: Excellent suggestion - will work it in.

File: rfcs/0000-rfc-session.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
+# 0000: Session
+<!--^ The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC, taking care not to conflict with other RFCs.-->
+
+- Stage: **0 (strawperson)** <!-- Update to reflect target stage -->
+- Date: 6/22/2020 <!-- Update to reflect date of most recent stage advancement -->
```
Comment (webmat) [inline rfcs/0000-rfc-session.md:5]: Note for the future: we set this at the time we merge the RFC. No need to twiddle with it :-)

No action is needed, though.

File: rfcs/0000-rfc-session.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+Who will be or has consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.
+
+e.g.:
+
+* @Yasmina | author
+* @Monique | sponsor
+* @EunJung | subject matter expert
+* @JaneDoe | grammar, spelling, prose
+* @Mariana
+-->
```
Comment (webmat) [inline rfcs/0000-rfc-session.md:205]: We're adding a section at the end of RFCs to add various related links, and also for RFCs to link to their PRs.

When you make the requested adjustments, could you also add this section at the end of the RFC, please?

```suggestion
-->


## References

<!-- Insert any links appropriate to this RFC in this section. -->

### RFC Pull Requests

<!-- An RFC should link to the PRs for each of it stage advancements. -->

* Stage 0: https://github.com/elastic/ecs/pull/879

<!--
* Stage 1: https://github.com/elastic/ecs/pull/NNN
...
-->
```

**Comment by jonathan-buttner** [comment]
Sorry I've been MIA! We've been swamped with the 7.9 release. @epixa @webmat other than what Mat outlined is there anything else I can help with to get this merged and moved to stage 1?

There are a couple people within the security org that I'd like to chime in on this as we iterate on the exact fields we'll be adding to ECS. Is stage 1 the appropriate time to do that?

**Comment by webmat** [comment]
This is great, thanks for reaching out about the interest :-)

So Dain is the original author, but we can move to having multiple co-authors to get this moving faster. We're definitely looking forward to the Endpoint team's input on this.

For simplicity's sake, if the team wants to comment about this today, you're welcome to comment on this PR directly.

The bar for merging as stage 0 is already met (this is appropriate for ECS), so we'll merge as soon as the file is moved to the right directory.

After stage 0 is merged, you or someone from Endpoint could open the next PR to continue the process of fleshing this out. Also, seeing the interest from Endpoint, it sounds like the team would like to be the official sponsor for these fields? 👍 

And to answer your question, yes stage 1 will be a great time to start fleshing out the list of fields we'll need for sessions :-)

**Comment by andrewvc** [comment]
Some thoughts from Uptime:

1. Having `type` and `kind` is confusing. More verbosity would help here. IMHO stick with one and add a qualifier
2. These sessions seem very network oriented, have you considered browser sessions and how they'd apply here?
3. It seems like adding the layer 7 protocol would make sense as well.

**Comment by dainperkins** [comment]
@andrewvc 

> 1. Having `type` and `kind` is confusing. More verbosity would help here. IMHO stick with one and add a qualifier

Maybe using more discretely specific qualifiers would help (not sure what those are but I get the point)

> 2. These sessions seem very network oriented, have you considered browser sessions and how they'd apply here?

Definitely want to cover browser sessions, applications sessions (sql, etc), I just didn;t have any handy when I wrote the rfc

> 3. It seems like adding the layer 7 protocol would make sense as well.

Would be all set with e.g. existing ECS fields  (network.protocol, etc.)

File: rfcs/0000-rfc-session.md (hunk 10 above)
Comment (dainperkins) [reply rfcs/0000-rfc-session.md:183]: done

File: rfcs/0000-rfc-session.md (hunk 11 above)
Comment (dainperkins) [reply rfcs/0000-rfc-session.md:32]: updated

```
These fields are used to track an entity's interaction with various assets, services, and applications in an enterprise.  Sessions will typically include a start event, often a login / authorization event performed locally or via network based mechanisms, and an end event indicating a logoff or session termination.  

When available, event start/end or duration fields should be populated, as well as iam, user, network, host, observer, process, source, destination, client, and server fields as appropriate to describe the specifics of the interaction.
```

File: rfcs/0000-rfc-session.md (hunk 9 above)
Comment (dainperkins) [reply rfcs/0000-rfc-session.md]: fixed

File: rfcs/0000-rfc-session.md (hunk 10 above)
Comment (dainperkins) [reply rfcs/0000-rfc-session.md]: fixed

**Comment by dainperkins** [comment]
@webmat 
Made a bunch of updates, added a text directory and moved the rfc md file there.  I think I need an approval after the changes

**Comment by ebeahan** [comment]
Thanks @dainperkins, and ++ to all the great discussion from everyone as well.

We'll assign the RFC number, update the date, and merge this in for stage 0.

**Comment by webmat** [comment]
@andrewvc Yes the goal for capturing sessions in ECS is to be inclusive of all the various kinds of sessions. Hopefully we can define fields that are applicable to all kinds of sessions. If need be, we can have fields specific to some kinds of sessions as well.

I'm envisioning this supporting:

* physical user sessions at a desktop
* various kinds of network sessions, including browser sessions

We can continue refining this on subsequent stage PRs

**Comment by joshdevins** [comment]
Did this make it into the v1.6.0 release? I don't see it in [the release notes](https://github.com/elastic/ecs/releases/tag/v1.6.0).

**Pull request merged**